### ISSUE TITLE
random MPI_Init() failure on ARM64 cpu (graviton 2) 

### ISSUE THREAD

**Issue by wzamazon** [open-mpi/ompi#8274]
Thank you for taking the time to submit an issue!

## Background information
When running Open MPI on graviton 2 cpu (ARM based), open mpi would randomly fail with the following error message:

```
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_interlib_declare
  --> Returned "Error" (-1) instead of "Success" (0)
--------------------------------------------------------------------------
```
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

v4.0.5

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

compiled from source, using option ```--with-sge --without-verbs  --with-libfabric=/path/to/libfabric "LDFLAGS=-Wl,--build-id -Wl,--enable-new-dtags"```

### Please describe the system on which you are running

* Operating system/version: 
Amazon Linux 2
* Computer hardware:
graviton 2  cpu (ARM 64) 
* Network type: 
EFA
-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

I was running some basic osu microbenchmark tests on a system equipped with AWS graviton 2 cpu using Open MPI, and I randomly encounter the following error:

The following is the command line I used
```mpirun --prefix /opt/amazon/openmpi \
        -n 256 -N 64 --tag-output  --hostfile <path_to_hostfile> \
        /path/to/omb/libexec/osu-micro-benchmarks/mpi/collective/osu_ibcast -x 1 -i 1 -m 4:8 
```

The following is the error message I got:
```
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_interlib_declare
  --> Returned "Error" (-1) instead of "Success" (0)
--------------------------------------------------------------------------
[ip-172-31-34-241:26268] *** An error occurred in MPI_Init
[ip-172-31-34-241:26268] *** reported by process [60817409,157]
[ip-172-31-34-241:26268] *** on a NULL communicator
[ip-172-31-34-241:26268] *** Unknown error
[ip-172-31-34-241:26268] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[ip-172-31-34-241:26268] ***    and potentially your MPI job)
[warn] Epoll MOD(1) on fd 65 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 65 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 100 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 100 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 101 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 101 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 76 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 76 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 164 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 164 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 74 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 74 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 73 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 73 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 106 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 106 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 63 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 63 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 110 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 110 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 157 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
[warn] Epoll MOD(4) on fd 157 failed.  Old events were 6; read change was 2 (del); write change was 0 (none): Bad file descriptor
[warn] Epoll MOD(1) on fd 192 failed.  Old events were 6; read change was 0 (none); write change was 2 (del): Bad file descriptor
.....
```

**Comment by wzamazon** [comment]
I did some debugging and the following is what I have found so far:

1. The error happened inside function ```ompi_interlib_declare()```, which is in file ```ompi/interlib/interlib.c```.

The following is a snippet related to this issue:

```
    .....
     opal_pmix.register_evhandler(&info, &directives, model_callback,
                                 model_registration_callback,
                                 (void*)&trk);
    OMPI_LAZY_WAIT_FOR_COMPLETION(trk.active);

    OPAL_LIST_DESTRUCT(&directives);
    OPAL_LIST_DESTRUCT(&info);
    if (OPAL_SUCCESS != trk.status) {
        return trk.status;
    }
    ....
```

My understanding is: ompi try to register an event handler ```model_callback``` with pmix for an event. It also provided a callback function for the registration's completion: ```model_registration_callback```.

The issue is with ```model_registration_callback```, whose code is:
```
static void model_registration_callback(int status,
                                        size_t errhandler_ref,
                                        void *cbdata)
{
    myreg_t *trk = (myreg_t*)cbdata;

    trk->status = status;
    interlibhandler_id = errhandler_ref;
    trk->active = false;
}
```
My understanding is: in the normal code path, pmix register the callback ```model_callback``` then call ```model_registraion_callback``` to notify ompi that registration is completed.

In OMPI, ```trk->status``` is initialized as OPAL_ERROR (the value is -1), and the expectation is that ```model_registration_callback``` will first do

```
   trk->status = status;
```

then do 

```
   trk->active = false;
```

Therefore when  ```OMPI_LAZY_WAIT_FOR_COMPLETION(trk.active);``` finished, ```trk->status``` has been set to 0. (if no error encountered inside pmix)

However, it seems that for ARM cpu, there is no guarantee that 

```
   trk->status = status;
```
happen before

```
   trk->active = false;
```

Hence, when ```OMPI_LAZY_WAIT_FOR_COMPLETION(trk.active);``` finish, ```trk->status``` sometime has not been updated (thus is still -1), which caused the error I see, e.g.

```
ompi_interlib_declare
  --> Returned "Error" (-1) instead of "Success" (0)
```

I tried to add a memory barrier 
```
asm volatile("" ::: "memory");
```
after 

```trk->status = status```.

With this change, I was able to run osu_ibcast 1000 times without hitting the issue.

I wonder if the above analysis make sense to you. e.g. the weaker memory ordering guarantee caused the issue I see?

**Comment by ggouaillardet** [comment]
Thanks for the report and analysis!

It makes perfect sense to me: ARM has a weak memory model (aka weakly consistent memory model aka relaxed memory model).

I believe the right way to do this in Open MPI is to call the "portable" `opal_atomic_mb()` (memory barrier), or maybe `opal_atomic_wmb()` (write memory barrier).

You are very welcome to issue a PR (the commit must be signed-off, please make sure you understand the implications) and tag @hjelmn.

**Comment by wzamazon** [comment]
@ggouaillardet

Thank you!  I will open PR for this.

**Comment by wzamazon** [comment]
Opened https://github.com/open-mpi/ompi/pull/8275.

Two more findings:
1. I found another place which has the same issue, and fixed it too.
2. master branch does not have this issue because it uses OPAL_PMIX_WAKEUP_THREAD instead of ```trk->active```.

**Comment by bosilca** [comment]
This is really unexpected in a single threaded execution, because the memory ordering should not have an impact there. Is this happening between multiple threads?

**Comment by wzamazon** [comment]
I am not too familiar with the code, but my understanding is that this is between two threads.

**Comment by rhc54** [comment]
Yes, it is between two threads - the upcall from the PMIx progress thread and the transfer to the main OPAL thread.

**Comment by wzamazon** [comment]
PR has been merged to both v4.1.x and v4.0.x branch. Closing ...
